### ISSUE TITLE
rules/python: Option to generate .pyc files for py_library.

### ISSUE THREAD

**Issue by adam-azarchs** [bazelbuild/bazel#16245]
### Description of the feature request:

Add options on one or more of the `py_library` rule itself, the python toolchain configuration, or the bazel command line, to generate python bytecode files as part of the build.

In addition to the basic on/off option, it might be useful to include options to
1. Control the "optimization" level (for what that's worth)
2. Leave the source `.py` file out of the build outputs - the `.pyc`-only use case.

### What underlying problem are you trying to solve with this feature?

When python loads a module, it parses the source code to an AST, which it then attempts to cache in bytecode format.  Building those as part of the build process solves a few issues:

1. To quote the documentation for the [py_compile](https://docs.python.org/3/library/py_compile.html) module:

    > Though not often needed, this function can be useful when installing modules for shared use, especially if some of the users may not have permission to write the byte-code cache files in the directory containing the source code.

    Particularly in the case where bazel is being used to build a tarball of code that includes (but may not be limited to) python, and might then be deployed somewhere that's read-only to most users, it would be useful to be able to include these precompiled bytecode files.
2. The _attempt_ to compile the bytecode files would fail on syntactically-invalid python code, which is probably a good thing for catching failures earlier on in the build process.
3. Having `.pyc` files available improves application startup time.   Especially for large python codebases, if some module is transitively imported from thousands of unit tests, currently each of those tests would end up re-parsing the python source file, which is a waste of time.  Having the `.pyc` files is also helpful for improving startup times for "serverless" platforms.
4. The `.pyc` files can be substantially smaller than the source files.  For situations where application distribution size is important, e.g. "serverless" platforms, this can matter.
5. Some people place value on the marginal degree of obfuscation and tamper resistance offered by `.pyc`-only distributions.  While reverse-engineering the source from a `.pyc` file isn't hard, it's also not nothing.

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

release 5.3.0

### Any other information, logs, or outputs that you want to share?

References/notes:
https://docs.python.org/3/library/py_compile.html
The default compilation mode is `TIMESTAMP`; this is probably a bad idea.  In a bazel build we'd probably want to use [`UNCHECKED_HASH`](https://docs.python.org/3/library/py_compile.html#py_compile.PycInvalidationMode.UNCHECKED_HASH).  Would also need to ensure that the embedded path name was appropriately relative.

From [PEP-3147](https://peps.python.org/pep-3147):
> Python will still support pyc-only distributions, however it will only do so when the pyc file lives in the directory where the py file would have been, i.e. not in the __pycache__ directory. pyc file outside of __pycache__ will only be imported if the py source file is missing.

This means that in the case where the `.py` file is still being included, the output path would need to depend on the python interpreter version.  This probably would require an attribute to be added to `py_runtime` for that purpose.

**Comment by rickeylev** [comment]
This is probably fairly easy to do by adding a `precompiler` attribute to the toolchain that the rules can use.

The tricky part I'm not sure of is what that attribute points to. It can't point to a py_binary itself, since that would be a circular dependency. But we have to use Python to run py_compile _somehow_.

Maybe allow it to be an executable or a single py file? If it's an executable, run it. If it's a py file, then run "$interpreter $file". This allows creating a pre-built executable if desired, while also allowing it to Just Work.

(this again touches on the problem of how there's several cases where py-binaries want to be used as part of the toolchain)

> In a bazel build we'd probably want to use UNCHECKED_HASH

Probably, yes. CHECKED_HASH is somewhat appealing, too, though, because it allows you to modify the source file and see the new behavior _without_ having to rebuild, though.

I'm +1 on defaulting to one and having an option for the other. Maybe base the default on `-c opt` so optimized builds used unchecked_hash by default, while fastbuilds use CHECKED_HASH by default. I can easily see a server not caring about this difference, while a CLI app would.

I also like the idea of a library-level control here somehow, too. A scenario we've enountered with large packages (e.g. tensorflow), is the sheer number of modules has its own overhead; pyc helps, but being as how modifying them is the exception, not the rule, making them check their hash is wasted work. We'll have to figure out how a target-level and global-flag-level option should interact, though (which has precedence?).

> This means that in the case where the .py file is still being included, the output path would need to depend on the python interpreter version. This probably would require an attribute to be added to py_runtime for that purpose.

Yeah, the details of the pyc infix aren't well understood to me. I think [PEP 488](https://peps.python.org/pep-0488/) is another half of the puzzle. I swear I thought there was a part of the specification that included the "compile mode" of Python itself, too (e.g. debug for non-debug), but I don't see that mentioned.

We also have to make sure that the generated name is recognized by the runtime that will be used. I think this is controlled by [BYTECODE_SUFFIXES](https://docs.python.org/3/library/importlib.html#importlib.machinery.BYTECODE_SUFFIXES)? One option is to simply stick this onto the toolchain, too. Another would be to pass the necessary suffix to the stub template so it can configure the runtime at startup.

Additionally, the output path depends on if the original .py files is included or not.

The Major.Minor Python version is in the `--python_version` flag.

**Comment by adam-azarchs** [comment]
> The tricky part I'm not sure of is what that attribute points to. It can't point to a py_binary itself, since that would be a circular dependency. But we have to use Python to run py_compile _somehow_.
>
> Maybe allow it to be an executable or a single py file? If it's an executable, run it. If it's a py file, then run "$interpreter $file". This allows creating a pre-built executable if desired, while also allowing it to Just Work.
> 
> (this again touches on the problem of how there's several cases where py-binaries want to be used as part of the toolchain)

Yeah, this is kind of a similar problem as the one faced for the coverage tool in #15590.  However, there's an important difference: `py_compile` is part of the python standard library.  Given that, I don't really see a need to offer an option to customization of the compilation tool - `bazel` could just hard-code running `$interpreter -m py_compile`, which as far as I can tell would work for all of CPython, PyPy, and MicroPython at least (though I've only tried with CPython).

> Probably, yes. CHECKED_HASH is somewhat appealing, too, though, because it allows you to modify the source file and see the new behavior _without_ having to rebuild, though.
> 
> I'm +1 on defaulting to one and having an option for the other. Maybe base the default on `-c opt` so optimized builds used unchecked_hash by default, while fastbuilds use CHECKED_HASH by default. I can easily see a server not caring about this difference, while a CLI app would.

Absolutely!  However, that's something you can handle with `select`, so it doesn't really have to impact the design much.

I think the options we'd want exposed look something like
1. No compilation.
2. `CHECKED_HASH`
3. `UNCHECKED_HASH`
4. `pyc`-only

As well as the mostly orthogonal choices for "optimization" level. (which I will continue to put in quotes. It's really too bad that there isn't an optimization level which strips out docstrings, which you don't need in a server context, but leaves in assertions, which might be a bit dangerous to strip out)

> I also like the idea of a library-level control here somehow, too. A scenario we've enountered with large packages (e.g. tensorflow), is the sheer number of modules has its own overhead; pyc helps, but being as how modifying them is the exception, not the rule, making them check their hash is wasted work. We'll have to figure out how a target-level and global-flag-level option should interact, though (which has precedence?).

This is why I suggested pre-compilation should be an attribute on the `py_library` rule, as well as possibly the ability to set a default at the toolchain level.  For one thing, you might want to have some things be distributed as pyc-only, while others  might want to include source.  You might also want to be able to control the "optimization" level on a per-target basis.  I think the safe approach there would be to just say local specification takes precedence over global specification - I can't really imagine a situation where someone would set an override on a library and we'd _not_ want to trust that they had a good reason to do so.  At a nitty-gritty API level, however, it does mean we have to be careful about the value of the attribute one sets for this; one might think that `compile_mode = None` would be a default of "use the toolchain default," but that would catch someone by surprise if they (somewhat reasonably) read that as meaning "no compilation".  Easy enough to handle - just make the default for the attribute be e.g. `"from_toolchain"` or whatever.

**Comment by rickeylev** [comment]
> Yeah, this is kind of a similar problem as the one faced for the coverage tool in #15590. However, there's an important difference: `py_compile` is part of the python standard library. Given that, I don't really see a need to offer an option to customization of the compilation tool - `bazel` could just hard-code running `$interpreter -m py_compile`, which as far as I can tell would work for all of CPython, PyPy, and MicroPython at least (though I've only tried with CPython).

I think invoking the interpreter directly like that is a workable intermediate solution. So I'm +1 on doing that for now. I'm fairly certain we need a custom program to ensure deterministic behavior and pass the necessary info, though. The cli of -m py_compile isn't sufficient.

However, I need to make a few architectural points about why such solutions make it hard or impossible to take full advantage of Bazel's processing model and features.

As some concrete examples:

When using embedded Python, there's no need for a standalone interpreter, so building one simply for precompilation is unnecessary work. (This is particularly salient to me because that's my primary way of building Python programs).

Precompilation is a near perfect fit for using persistent workers, which would eliminate runtime startup overhead. This requires special support by the executable invoked.

Precompilation must run in the exec configuration. This means the runtime itself may be built differently than the runtime actually used, may be run remotely, and/or run on an entirely different platform. So more steps have to be taken to ensure such details don't leak into the output. For example, the launcher script does several things to prevent system settings from affecting the runtime which have to be re-implemented.

Similarly, the exec configuration usually emphasizes time-to-build over runtime-efficiency. This makes sense for typical user build tools, but less so for something like precompilation; precompilation is well defined and it's rate of invocation scales with the size of the overall build, ie the larger your transitive closure, the more it runs, so you're probably going to benefit from the precompiler (and runtime) being built with optimizations enabled. Such optimizations can extend beyond simply e.g. pyc -O levels, and into C-level optimizations that affect the runtime itself; these optimizations are hard to apply during a build, but are easy to apply to to an prebuilt executable which is later run.

This isn't to say simply doing "$runtime -m py_compile" is _bad_ or _wrong_, just that it leaves a lot sitting at the Bazel Buffet Table.

> As well as the mostly orthogonal choices for "optimization" level. (which I will continue to put in quotes. It's really too bad that there isn't an optimization level which strips out docstrings, which you don't need in a server context, but leaves in assertions, which might be a bit dangerous to strip out)
> 

Srsly, right? This has bugged me for years! It's the sort of thing a custom precompiler executable _could_ do, though. This also makes me think, if we exposed an optimization level, it'd probably be better done as a list of values instead of a simple number.

> I think the safe approach there would be to just say local specification takes precedence over global specification - I can't really imagine a situation where someone would set an override on a library and we'd _not_ want to trust that they had a good reason to do so.

I agree. A case we have internally is that protocol buffers (which are handled by a special rule) are compiled by default. This is pretty akin to a library-level setting.

> At a nitty-gritty API level, however, it does mean we have to be careful about the value of the attribute one sets for this; one might think that `compile_mode = None` would be a default of "use the toolchain default," but that would catch someone by surprise if they (somewhat reasonably) read that as meaning "no compilation". Easy enough to handle - just make the default for the attribute be e.g. `"from_toolchain"` or whatever.

I agree.

**Comment by adam-azarchs** [comment]
Worth noting mostly for the sake of code archeology that there's this comment in the code currently: https://github.com/bazelbuild/bazel/blob/6d72ca979f8cf582f53452d5f905346e7effb113/src/main/java/com/google/devtools/build/lib/rules/python/PythonSemantics.java#L69-L71
however as far as I can tell the implementation https://github.com/bazelbuild/bazel/blob/9d0163002ca63f4cdbaff9420380d72e2a6e38b3/src/main/java/com/google/devtools/build/lib/bazel/rules/python/BazelPythonSemantics.java#L117-L120
simply copy the sources unaltered. But for what it's worth I'm guessing that at some point in the past someone had an implementation, or at least and attempt at one, probably Google-internal-only, for at least some of this functionality.

**Comment by Ryang20718** [comment]
I would very much <3 to see this feature. 

<details>

<summary>we have a lot of heavy imports :( </summary>

Some of our heavy hitter imports bog down both test & runtime (mainly test)

One of the pain points right now is that we use pytorch in our bazel monorepo. However, because we import it in so many of our libraries, it incurs a large import time overhead (3.8 seconds).

With pycache, it drops down to ~2.5 seconds.
</details>


Is the general idea to generate pyc files once when all external third party repos are fetched at analysis time and then to ingest these files as input in `py_library`?

Would love some pointers on where to start (even if it's just a hacky prototype 😅 )

**Comment by rickeylev** [comment]
I'm going to transfer this issue to the rules_python repo because, as of rules_python 0.31.0, we've enabled the rules_python-based implementation of the rules by default for Bazel 7+. This basically means this feature won't land in the Bazel builtin rules, but would instead be in the rules_python implementation.
